components/eam/src/chemistry/aerosol/wetdep.F90

```diff
@@ -502,798 +471,732 @@ subroutine wetdepa_v2( ncol, deltat, &
-!      omsm = 1.-1.e-10          ! used to prevent roundoff errors below zero
-      omsm = 1._r8-2*epsilon(1._r8) ! used to prevent roundoff errors below zero
-      precmin =  0.1_r8/8.64e4_r8      ! set critical value to 0.1 mm/day in kg/m2/s
-
-      adjfac = deltat/(max(deltat,cmftau)) ! adjustment factor from hack scheme
-
-      ! assume 4 m/s fall speed currently (should be improved)
-!      vfall = 4.
-	
-      ! default (if other sol_facts aren't in call, set all to required sol_fact
-!     sol_facti = sol_fact
-!     sol_factb = sol_fact
-!     sol_factii = sol_fact
-!     sol_factbi = sol_fact
-
-!     if ( present(sol_facti_in) )  sol_facti = sol_facti_in
-!     if ( present(sol_factii_in) )  sol_factii = sol_factii_in
-!     if ( present(sol_factbi_in) )  sol_factbi = sol_factbi_in
-!     sol_facti = sol_facti_in
-!     sol_factii = sol_factii_in
-!     sol_factbi = sol_factbi_in
-
-!     sol_factic  = sol_facti
-!     sol_factiic = sol_factii
-!     if ( present(sol_factic_in ) )  sol_factic  = sol_factic_in
-!     if ( present(sol_factiic_in) )  sol_factiic = sol_factiic_in
-!     sol_factic  = sol_factic_in
-!     sol_factiic = sol_factiic_in
 
-      ! this section of code is for highly soluble aerosols,
-      ! the assumption is that within the cloud that
-      ! all the tracer is in the cloud water
-      !
-      ! for both convective and stratiform clouds, 
-      ! the fraction of cloud water converted to precip defines
-      ! the amount of tracer which is pulled out.
+main_i_loop: &
+      do icol = 1,ncol
 
-      do i = 1,pcols
-         precabs(i) = 0
-         precabc(i) = 0
-         scavab(i) = 0
-         scavabc(i) = 0
-         tracab(i) = 0
-         cldmabs(i) = 0
-         cldmabc(i) = 0
-
-         precabs_base(i) = 0.0_r8
-         precabc_base(i) = 0.0_r8
-         precnums_base(i) = 0.0_r8
-         precnumc_base(i) = 0.0_r8
-         resusp_c_sv(i) = 0.0_r8
-         resusp_s_sv(i) = 0.0_r8
-
-       ! Jan.16. Sungsu 
-       ! I added below to compute vertically projected cumulus and stratus fractions from the top to the
-       ! current model layer by assuming a simple independent maximum overlapping assumption for 
-       ! each cloud.
-       ! cldovr_cu(i) = 0._r8
-       ! cldovr_st(i) = 0._r8
-       ! End by Sungsu
+         ! initiate variables
+         precabs = 0.0_r8
+         precabc = 0.0_r8
+         scavabs = 0.0_r8
+         scavabc = 0.0_r8
 
-      end do
+         precabs_base = 0.0_r8
+         precabc_base = 0.0_r8
+         precnums_base = 0.0_r8
+         precnumc_base = 0.0_r8
 
 main_k_loop: &
-      do k = 1,pver
-main_i_loop: &
-         do i = 1,ncol
-            tc     = t(i,k) - tmelt
-            weight = max(0._r8,min(-tc*0.05_r8,1.0_r8)) ! fraction of condensate that is ice
-            weight = 0._r8                                 ! assume no ice
-
-            pdog = pdel(i,k)/gravit
+         do kk = 1,pver
```

Comment:
I switch the order of the nested loops because many of the temporary variables (such as precabc) are column variables and have only a dimension on icol. It looks more logical to me that integrate through the entire vertical layer (kk) within each icol, and the subroutines within these loops can have inputs and outputs for single-value variables only, otherwise the subroutines will have inputs/outputs of 1-d arrays (icol). Is there a way to test the speed of the loop?

Reply:
I created a quick and dirty program that demonstrates that row vs. column-major data accesses make a difference in the timings. [Here](https://gist.github.com/singhbalwinder/aaa55edebe909ffeb7ca23eef4f1c025) is the program.

We can test it within E3SM but the difference may be sensitive to the model resolution. It is better to switch the ordering as it is a known fact. I will read the code once more to see if there is a better way to reorganize the code.

Reply:
Sorry, forgot to include the output. Here is what I get on Compy:
```
Row major:   1.238548    
 Column major:  0.2171589  
```

Reply:
I made a test with subroutine wetdepa_v2 by assigning random inputs with pcols=10000 and pver=1000. it does make differences (5.846s vs 9.321s). I'll change the loop back to column major.